### DISCUSSION TITLE
Feedback on 7 rc.11

### DISCUSSION THREAD

**Discussion by ngbrown** [rabbitmq/rabbitmq-dotnet-client#1694] [Q&A]
@lukebakken had asked me for feedback on the v7 RC, so this is my initial feedback.

I updated the `rabbit-rpc` branch on the repo here: https://github.com/ngbrown/grpc-testing/commits/rabbit-rpc/, so look at the recent commits.

### Describe the solution you'd like

- The conversion of methods to Async went ok because the method names changed. If there are any that the name didn't change and also converted to async, then I probably missed them.
- I'm not clear on what value of `mandatory` is equivalent to the v6 behavior. The parameter also isn't optional, at least on the `BasicPublishAsync` method I use. This parameter should have a default value for easier migration. Or at least mention in the Doc XML what the equivalent value is for v6 behavior.

### Additional context

I also plan on testing how `System.Diagnostics.Activity` works with `AsyncEventingBasicConsumer.Received` event handlers, but I won't be able to look at that until the end of the week, so thought that I would get this feedback out there.

**Comment by michaelklishin** [comment]
We should make `mandatory` default to `false`, it is the case in other clients and in general, [mandatory publishing](https://www.rabbitmq.com/docs/publishers#amqp-0-9-1-2) is not what the majority of users need, most of the time.

**Comment by lukebakken** [comment]
> I'm not clear on what value of mandatory is equivalent to the v6 behavior. The parameter also isn't optional, at least on the BasicPublishAsync method I use.

Which method do you use? Can you point it out in the 6.x code? I'm guessing it is one of these:

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/api/IModelExtensions.cs#L77-L105

**Reply by lukebakken** [reply]
Unfortunately I can't seem to restore the v6 behavior without causing "ambiguous reference" errors in the code.

**Reply by ngbrown** [reply]
In v6, I was using:

```cs
this._channel.BasicPublish(exchange: this._replyExchangeName, routingKey: this._props.ReplyTo, basicProperties: replyProps, body: responseBytes)
```

**Reply by lukebakken** [reply]
Thanks, I'll check again but there's a good chance we can't support this particular overload.

**Reply by ngbrown** [reply]
So all my calls to `BasicPublish` went to https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/86a7b0fb8c0a4980722aee2611696762f903fa89/projects/RabbitMQ.Client/client/api/IModelExtensions.cs#L94

I see by looking at the other overloads in v6 that `mandatory = false` was the default. My suggestion would be to at least add that default information to the Doc XML.

**Comment by ngbrown** [comment]
I started working on integrating with the `ActivitySource` addition in the v7 RabbitMQ Client. There doesn't appear to be any `DiagnosticListener` I can subscribe to. I'm using `SerilogTracing` and it would be ideal to be able to subscribe to the RabbitMQ activities and generate my own templated strings that are not unique on every activity. This would eventually be rolled into a library like is available [here](https://github.com/serilog-tracing/serilog-tracing/blob/dev/src/SerilogTracing.Instrumentation.SqlClient/Instrumentation/SqlClient/SqlCommandActivityInstrumentor.cs).

I'm getting the activities from RabbitMQ fine, but the only two diagnostic sources are "Microsoft.Extensions.Hosting" and "Microsoft.AspNetCore".

From the [Activity User Guide](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md):

> DO - create new DiagnosticListener for specific Activity type to allow filtering by activity. E.g Incoming and Outgoing Http Requests should be different DiagnosticListeners. Follow [DiagnosticSource User Guide](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md) to pick a name.

I should be able to filter on the RabbitMQ activities by using the `DiagnosticListener.AllListeners.Subscribe` API.

**Reply by lukebakken** [reply]
> There doesn't appear to be any DiagnosticListener I can subscribe to

Is this something that exists in the 6.x version of this client?

**Reply by ngbrown** [reply]
No, this is a new feature of the v7 client via #1261.

**Reply by lukebakken** [reply]
Ah, gotcha. Well, we've only used ActivitySource to add support for OpenTelemetry, and to be honest, I'm not very familiar with how that works.

Maybe @stebet can chime in?
